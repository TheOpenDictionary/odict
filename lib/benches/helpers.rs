use std::sync::LazyLock;

use odict::{DictionaryFile, DictionaryReader, DictionaryWriter, Result};

pub fn create_archive_dict(name: &str) -> Result<DictionaryFile> {
    let reader = DictionaryReader::default();
    let writer = DictionaryWriter::default();
    let input = format!("../examples/{}.xml", name);
    let output = format!("../examples/{}.odict", name);

    writer.compile_xml(&input, &output)?;
    reader.read_from_path(&output)
}

pub static EXAMPLE_DICTIONARY_1: LazyLock<DictionaryFile> =
    LazyLock::new(|| create_archive_dict("wiktionary").unwrap());
