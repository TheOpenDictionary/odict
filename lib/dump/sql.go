package dump

import (
	"bytes"
	"embed"

	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/bokwoon95/sq"
	"github.com/bokwoon95/sqddl/ddl"
)

//go:embed sql.go
var fsys embed.FS

var sqlDictionaryId int = 1
var sqlEntryId int = 1
var sqlEtymologyId int = 1
var sqlSenseId int = 1
var sqlGroupId int = 1
var sqlDefinitionId int = 1
var sqlExampleId int = 1

func sql(dict types.DictionaryRepresentable, sqlDialect SqlDialect) (string, error) {
	var sqlCmds string

	create, createErr := sqlCreate(sqlDialect)
	insert, insertErr := sqlInsert(sqlDialect, dict)

	if createErr != nil {
		return "", createErr
	} else {
		sqlCmds += create
	}

	if insertErr != nil {
		return "", insertErr
	}

	sqlCmds += insert

	return sqlCmds, nil
}

func sqlCreate(sqlDialect SqlDialect) (string, error) {
	var buf bytes.Buffer

	// Generate SQL create statements and constraints
	generateCmd := &ddl.GenerateCmd{
		Dialect:   sqlDialect,
		DirFS:     fsys,
		Filenames: []string{"sql.go"},
		DryRun:    true,
		Stdout:    &buf,
	}

	err := generateCmd.Run()

	if err != nil {
		return "", err
	}

	return buf.String(), nil
}

func sqlInsert(sqlDialect SqlDialect, dict types.DictionaryRepresentable) (string, error) {
	var sqlCmds string

	// Entry point that will capture every insert statement
	// Dictionary, entries, etymologies, senses, groups, definitions, examples
	insert, err := sqlInsertDictionary(sqlDialect, dict)

	if err != nil {
		return "", err
	}

	sqlCmds += insert

	return sqlCmds, nil
}

func sqlInsertDictionary(sqlDialect SqlDialect, dict types.DictionaryRepresentable) (string, error) {
	var sqlCmds string

	// Insert dictionary w/ PK of 1
	d := sq.New[DICTIONARIES]("")

	insertQuery := sq.
		InsertInto(d).
		Columns(d.NAME, d.ID).
		Values(sq.Literal(dict.Name), sq.Literal(sqlDictionaryId))

	dict_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

	if err != nil {
		return "", err
	}

	sqlCmds += dict_query + ";\n"

	insertEntries, insertErr := sqlInsertEntries(sqlDialect, dict.Entries)

	if insertErr != nil {
		return "", insertErr
	}

	sqlCmds += insertEntries

	return sqlCmds, nil
}

func sqlInsertEntries(sqlDialect SqlDialect, entries types.KVMap[string, types.EntryRepresentable]) (string, error) {
	var sqlCmds string

	// Insert entries w/ relation to dictionary with PK of 1
	for _, entry := range entries {
		e := sq.New[ENTRIES]("")

		insertQuery := sq.
			InsertInto(e).
			Columns(e.ID, e.TERM, e.DICTIONARY_ID).
			Values(sq.Literal(sqlEntryId), sq.Literal(entry.Term), sq.Literal(1))

		e_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

		if err != nil {
			return "", err
		}

		sqlCmds += e_query + ";\n"

		insertEty, err := sqlInsertEtymologies(sqlDialect, entry.Etymologies)

		if err != nil {
			return "", err
		}

		sqlCmds += insertEty
		sqlEntryId++
	}

	return sqlCmds, nil
}

func sqlInsertEtymologies(sqlDialect SqlDialect, etymologies []types.EtymologyRepresentable) (string, error) {
	var sqlCmds string

	// Insert etymologies w/ relation to current entry
	for _, etymology := range etymologies {
		ety := sq.New[ETYMOLOGIES]("")

		insertQuery := sq.
			InsertInto(ety).
			Columns(ety.ID, ety.DESCRIPTION, ety.ENTRY_ID).
			Values(sq.Literal(sqlEtymologyId), sq.Literal(etymology.Description), sq.Literal(sqlEntryId))

		ety_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

		if err != nil {
			return "", err
		}

		sqlCmds += ety_query + ";\n"

		insertSenses, insertErr := sqlInsertSenses(sqlDialect, etymology.Senses)

		if insertErr != nil {
			return "", insertErr
		}

		sqlCmds += insertSenses
		sqlEtymologyId++
	}

	return sqlCmds, nil
}

func sqlInsertSenses(sqlDialect SqlDialect, senses types.KVMap[types.PartOfSpeech, types.SenseRepresentable]) (string, error) {
	var sqlCmds string

	// Insert senses w/ relation to current ety
	for _, sense := range senses {
		usg := sq.New[SENSES]("")

		insertQuery := sq.
			InsertInto(usg).
			Columns(usg.ID, usg.ETYMOLOGY_ID).
			Values(sq.Literal(sqlSenseId), sq.Literal(sqlEtymologyId))

		usg_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

		if err != nil {
			return "", err
		}

		sqlCmds += usg_query + ";\n"

		insertGroups, insertErr := sqlInsertGroups(sqlDialect, sense.Groups)

		if insertErr != nil {
			return "", insertErr
		}

		sqlCmds += insertGroups
		sqlSenseId++
	}

	return sqlCmds, nil
}

func sqlInsertGroups(sqlDialect SqlDialect, groups []types.GroupRepresentable) (string, error) {
	var sqlCmds string

	// Insert groups w/ relation to current sense
	for _, group := range groups {
		grp := sq.New[GROUPS]("")

		insertQuery := sq.
			InsertInto(grp).
			Columns(grp.ID, grp.DESCRIPTION, grp.SENSE_ID).
			Values(sq.Literal(sqlGroupId), sq.Literal(group.Description), sq.Literal(sqlSenseId))

		grp_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

		if err != nil {
			return "", err
		}

		insertDefinitions, insertErr := sqlInsertDefinitions(sqlDialect, group.Definitions)

		if insertErr != nil {
			return "", insertErr
		}

		sqlCmds += grp_query + ";\n"
		sqlCmds += insertDefinitions

		sqlGroupId++
	}

	return sqlCmds, nil
}

func sqlInsertDefinitions(sqlDialect SqlDialect, definitions []types.DefinitionRepresentable) (string, error) {
	var sqlCmds string

	// Insert definitions w/ relation to current sense/group
	for _, definition := range definitions {
		def := sq.New[DEFINITIONS]("")

		insertQuery := sq.
			InsertInto(def).
			Columns(def.ID, def.TEXT, def.SENSE_ID, def.GROUP_ID).
			Values(sq.Literal(sqlDefinitionId), sq.Literal(definition.Value), sq.Literal(sqlSenseId), sq.Literal(sqlGroupId))

		def_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

		if err != nil {
			return "", err
		}

		insertExamples, insertErr := sqlInsertExamples(sqlDialect, definition.Examples)

		if insertErr != nil {
			return "", insertErr
		}

		sqlCmds += def_query + ";\n"
		sqlCmds += insertExamples

		sqlDefinitionId++
	}

	return sqlCmds, nil
}

func sqlInsertExamples(sqlDialect SqlDialect, examples []string) (string, error) {
	var sqlCmds string

	// Insert examples w/ relation to current definition
	for _, example := range examples {
		ex := sq.New[EXAMPLES]("")

		insertQuery := sq.
			InsertInto(ex).
			Columns(ex.ID, ex.TEXT, ex.DEFINITION_ID).
			Values(sq.Literal(sqlExampleId), sq.Literal(example), sq.Literal(sqlDefinitionId))

		ex_query, _, err := sq.ToSQL(sqlDialect, insertQuery, nil)

		if err != nil {
			return "", err
		}

		sqlCmds += ex_query + ";\n"

		sqlExampleId++
	}

	return sqlCmds, nil
}
