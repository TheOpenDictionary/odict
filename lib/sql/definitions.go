package sql

import (
	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/bokwoon95/sq"
	"github.com/google/uuid"
)

type DEFINITIONS struct {
	sq.TableStruct
	ID       sq.UUIDField   `ddl:"primarykey"`
	VALUE    sq.StringField `ddl:"notnull"`
	SENSE_ID sq.UUIDField   `ddl:"references=senses.id"`
	GROUP_ID sq.UUIDField   `ddl:"references=groups.id"`
}

// Insert definitions w/ relation to current sense/group
func insertDefinitions(builder *SQLBuilder, senseID string, groupID string, definitions []types.DefinitionRepresentable) {
	for _, definition := range definitions {
		def := sq.New[DEFINITIONS]("")
		id := uuid.New().String()
		value := definition.Value.String()

		builder.AddCommand(
			sq.
				InsertInto(def).
				ColumnValues(func(col *sq.Column) {
					col.Set(def.ID, sq.Literal(id))
					col.Set(def.VALUE, sq.Literal(value))

					if len(senseID) > 0 {
						col.Set(def.SENSE_ID, sq.Literal(senseID))
					}

					if len(groupID) > 0 {
						col.Set(def.GROUP_ID, sq.Literal(groupID))
					}
				}),
		)

		insertNotes(builder, id, definition.Notes)
		insertExamples(builder, id, "", definition.Examples)
	}
}
