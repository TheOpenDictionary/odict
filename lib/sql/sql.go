package sql

import (
	"bytes"
	"embed"
	"strings"

	"github.com/TheOpenDictionary/odict/lib/types"
	"github.com/bokwoon95/sqddl/ddl"
)

//go:embed definitions.go dictionaries.go
//go:embed entries.go etymologies.go
//go:embed examples.go groups.go
//go:embed notes.go senses.go
var fsys embed.FS

func createSchema(dialect SQLDialect) (string, error) {
	var buf bytes.Buffer

	// Generate SQL create statements and constraints
	generate := &ddl.GenerateCmd{
		Dialect: dialect,
		DirFS:   fsys,
		Filenames: []string{
			"dictionaries.go",
			"entries.go",
			"etymologies.go",
			"senses.go",
			"groups.go",
			"definitions.go",
			"examples.go",
			"notes.go",
		},
		DryRun: true,
		Stdout: &buf,
	}

	err := generate.Run()

	if err != nil {
		return "", err
	}

	return buf.String(), nil
}

func DictionaryToSQL(dict types.DictionaryRepresentable, dialect SQLDialect, includeSchema bool) (string, error) {
	builder := createSQLBuilder(dialect)

	insertDictionary(&builder, dict)

	insert, err := builder.Build(dialect)

	if err != nil {
		return "", err
	}

	cmds := []string{insert}

	if includeSchema {
		schema, err := createSchema(dialect)

		if err != nil {
			return "", err
		}

		cmds = append([]string{schema}, cmds...)
	}

	return strings.Join(cmds, ";\n\n"), nil
}
