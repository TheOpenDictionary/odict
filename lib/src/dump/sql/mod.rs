mod definitions;
mod dictionaries;
mod entries;
mod etymologies;
mod examples;
mod groups;
mod notes;
mod senses;
mod sql;
mod utils;

pub use self::sql::*;
pub use self::utils::SQLDialect;
