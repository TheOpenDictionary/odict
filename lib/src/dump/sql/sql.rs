use std::error::Error;

use crate::{Dictionary, SQLDialect};

use super::{
    dictionaries::{create_dictionaries, insert_dictionary},
    utils::SQLBuilder,
};

pub trait ToSQL {
    fn to_sql(self, dialect: SQLDialect) -> Result<String, Box<dyn Error>>;
}

fn add_schema(builder: &mut SQLBuilder) {
    create_dictionaries(builder);
}

impl ToSQL for Dictionary {
    fn to_sql(self, dialect: SQLDialect) -> Result<String, Box<dyn Error>> {
        let mut builder = SQLBuilder::new(dialect);

        add_schema(&mut builder);

        insert_dictionary(&mut builder, self);

        builder.build()
    }
}
