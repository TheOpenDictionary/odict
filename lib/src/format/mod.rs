#[cfg(feature = "html")]
pub mod html;
#[cfg(feature = "json")]
pub mod json;
#[cfg(feature = "markdown")]
pub mod md;
#[cfg(feature = "sql")]
pub mod sql;
pub mod xml;
