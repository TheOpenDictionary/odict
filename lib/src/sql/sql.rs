use std::error::Error;

use crate::{sql::SQLDialect, Dictionary};

use super::{
    definitions::create_definitions,
    dictionaries::{create_dictionaries, insert_dictionary},
    entries::create_entries,
    etymologies::create_etymologies,
    examples::create_examples,
    groups::create_groups,
    notes::create_notes,
    senses::create_senses,
    utils::SQLBuilder,
};

pub trait ToSQL {
    fn to_sql(self, dialect: SQLDialect) -> Result<String, Box<dyn Error>>;
}

fn add_schema(builder: &mut SQLBuilder) {
    create_dictionaries(builder);
    create_entries(builder);
    create_etymologies(builder);
    create_senses(builder);
    create_groups(builder);
    create_definitions(builder);
    create_notes(builder);
    create_examples(builder);
}

impl ToSQL for Dictionary {
    fn to_sql(self, dialect: SQLDialect) -> Result<String, Box<dyn Error>> {
        let mut builder = SQLBuilder::new(dialect);

        add_schema(&mut builder);

        insert_dictionary(&mut builder, &self)?;

        builder.build()
    }
}
