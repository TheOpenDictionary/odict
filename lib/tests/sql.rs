mod helpers;

#[cfg(test)]
mod index_tests {

    use insta::assert_snapshot;
    use odict::format::sql::{SQLDialect, ToSQL};
    use regex::Regex;

    use crate::helpers::EXAMPLE_DICT_1;

    #[test]
    fn test_sql() {
        let result = EXAMPLE_DICT_1
            .to_dictionary()
            .unwrap()
            .clone()
            .to_sql(SQLDialect::Postgres);

        let re = Regex::new(
            r"[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}",
        )
        .unwrap();

        assert_eq!(result.is_err(), false);
        assert_snapshot!(re.replace_all(result.unwrap().as_str(), ""));
    }
}
