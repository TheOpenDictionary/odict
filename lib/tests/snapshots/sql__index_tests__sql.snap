---
source: lib/tests/sql.rs
expression: "re.replace_all(result.unwrap().as_str(), \"\")"
snapshot_kind: text
---
CREATE TABLE IF NOT EXISTS "dictionaries" ( "id" uuid NOT NULL PRIMARY KEY, "name" varchar );
CREATE TABLE IF NOT EXISTS "entries" ( "id" uuid NOT NULL PRIMARY KEY, "term" varchar NOT NULL, "dictionary_id" uuid NOT NULL, FOREIGN KEY ("dictionary_id") REFERENCES "dictionaries" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
CREATE TABLE IF NOT EXISTS "etymologies" ( "id" uuid NOT NULL PRIMARY KEY, "description" varchar, "entry_id" uuid NOT NULL, FOREIGN KEY ("entry_id") REFERENCES "entries" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
CREATE TABLE IF NOT EXISTS "senses" ( "id" uuid NOT NULL PRIMARY KEY, "pos" varchar, "etymology_id" uuid NOT NULL, FOREIGN KEY ("etymology_id") REFERENCES "etymologies" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
CREATE TABLE IF NOT EXISTS "groups" ( "id" uuid NOT NULL PRIMARY KEY, "description" varchar, "index" integer NOT NULL, "sense_id" uuid NOT NULL, FOREIGN KEY ("sense_id") REFERENCES "senses" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
CREATE TABLE IF NOT EXISTS "definitions" ( "id" uuid NOT NULL PRIMARY KEY, "value" varchar NOT NULL, "index" integer NOT NULL, "sense_id" uuid, "group_id" uuid, FOREIGN KEY ("sense_id") REFERENCES "senses" ("id") ON DELETE CASCADE ON UPDATE CASCADE, FOREIGN KEY ("group_id") REFERENCES "groups" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
CREATE TABLE IF NOT EXISTS "notes" ( "id" uuid NOT NULL PRIMARY KEY, "value" varchar, "index" integer NOT NULL, "definition_id" uuid NOT NULL, FOREIGN KEY ("definition_id") REFERENCES "definitions" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
CREATE TABLE IF NOT EXISTS "examples" ( "id" uuid NOT NULL PRIMARY KEY, "value" varchar, "index" integer NOT NULL, "definition_id" uuid, "note_id" uuid, FOREIGN KEY ("definition_id") REFERENCES "definitions" ("id") ON DELETE CASCADE ON UPDATE CASCADE, FOREIGN KEY ("definition_id") REFERENCES "notes" ("id") ON DELETE CASCADE ON UPDATE CASCADE );
INSERT INTO "dictionaries" ("id", "name") VALUES ('', 'Example Dictionary 1');
INSERT INTO "entries" ("id", "term", "dictionary_id") VALUES ('', 'cat', '');
INSERT INTO "etymologies" ("id", "description", "entry_id") VALUES ('', 'Latin root', '');
INSERT INTO "senses" ("id", "pos", "etymology_id") VALUES ('', 'noun', '');
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'a cat', 0, '', NULL);
INSERT INTO "notes" ("id", "value", "index", "definition_id") VALUES ('', 'Some definition note', 0, '');
INSERT INTO "examples" ("id", "value", "index", "definition_id", "note_id") VALUES ('', 'Some example', 0, NULL, '');
INSERT INTO "examples" ("id", "value", "index", "definition_id", "note_id") VALUES ('', 'There goes a cat!', 0, '', NULL);
INSERT INTO "entries" ("id", "term", "dictionary_id") VALUES ('', 'dog', '');
INSERT INTO "etymologies" ("id", "description", "entry_id") VALUES ('', 'Latin root', '');
INSERT INTO "senses" ("id", "pos", "etymology_id") VALUES ('', 'unknown', '');
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'a dog', 0, '', NULL);
INSERT INTO "entries" ("id", "term", "dictionary_id") VALUES ('', 'poo', '');
INSERT INTO "etymologies" ("id", "description", "entry_id") VALUES ('', 'Latin root', '');
INSERT INTO "senses" ("id", "pos", "etymology_id") VALUES ('', 'unknown', '');
INSERT INTO "groups" ("id", "description", "index", "sense_id") VALUES ('', 'A number of verb senses', 0, '');
INSERT INTO "entries" ("id", "term", "dictionary_id") VALUES ('', 'ran', '');
INSERT INTO "entries" ("id", "term", "dictionary_id") VALUES ('', 'run', '');
INSERT INTO "etymologies" ("id", "description", "entry_id") VALUES ('', 'Latin root', '');
INSERT INTO "senses" ("id", "pos", "etymology_id") VALUES ('', 'noun', '');
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', '(vertebrates) To move swiftly', 0, '', NULL);
INSERT INTO "examples" ("id", "value", "index", "definition_id", "note_id") VALUES ('', 'The dog runs after the cat.', 0, '', NULL);
INSERT INTO "examples" ("id", "value", "index", "definition_id", "note_id") VALUES ('', 'The horse ran away.', 1, '', NULL);
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip.', 1, '', NULL);
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'A pleasure trip.', 2, '', NULL);
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'Flight, instance or period of fleeing.', 3, '', NULL);
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'Migration (of fish).', 4, '', NULL);
INSERT INTO "definitions" ("id", "value", "index", "sense_id", "group_id") VALUES ('', 'A group of fish that migrate, or ascend a river for the purpose of spawning.', 5, '', NULL);
INSERT INTO "senses" ("id", "pos", "etymology_id") VALUES ('', 'verb', '');
INSERT INTO "groups" ("id", "description", "index", "sense_id") VALUES ('', 'A number of verb senses', 0, '')
